### PR TITLE
Adds new parser AliasDevice and fast search brands by indexes

### PR DESCRIPTION
**Pull request by sanchezzzhak**
issue #6267

### CODE REVIEW

**Comment by sanchezzzhak** [comment]
@sgiehl is there a ready setup PHPCS for phpstorm 2020.3 ?

**Comment by sgiehl** [comment]
if you installed the composer dependencies there should be one in `vendor\bin\phpcs` (and `vendor\bin\phpcbf` for automatic fixes)

File: Tests/Parser/fixtures/alias_devices.yml [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ alias:
+   name: L270
+
+-
+ user_agent: Mozilla/5.0 (Linux; U; Android 7.1.2; uz-Latn-UZ; U5 Build/N2G47H) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/57.0.2987.108 UCBrowser/12.10.0.1163 UCTurbo/1.9.3.900 Mobile Safari/537.36
+ alias:
+   name: U5
+
+-
+ user_agent: Mozilla/5.0 (Android 7.0; samsung SM-A710F) AppleWebKit/537.36 KHTML, like Gecko Version/4.0 Chrome/30.0.0.0 Mobile Safari/537.36 SurfBrowser/3.0
```
Comment (mimmi20) [inline Tests/Parser/fixtures/alias_devices.yml:37]: Please fix indentation

File: Tests/Parser/AliasDeviceTest.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        $result = $parser->parse();
+
+        if (\is_array($result)) {
+            $this->assertArrayHasKey('name', $result);
+            $this->assertEquals($alias, $result);
+
+            return;
+        }
+
+        $this->assertEquals(null, $result);
```
Comment (mimmi20) [inline Tests/Parser/AliasDeviceTest.php:41]: Please use `$this->assertNull($result)`

**Review by mimmi20 — CHANGES_REQUESTED** [review]

**Review by mimmi20 — DISMISSED** [review]

**Comment by sanchezzzhak** [comment]
I found a good use for AliasDevice, I will try to pass the changes from https://github.com/sanchezzzhak/node-device-detector/tree/pr_device_info8 in PHP

My 22k tests were reduced from 2 minutes to 43 seconds, without cache

**Comment by sanchezzzhak** [comment]
summary

Added new methods to device parsers
public list:
AbstractDeviceParser::setBrandIndexes(array brands)  - parsing only selected brands
AbstractDeviceParser::parseAllMatch()    - same as parse, only returns all matches
some checks in the code have been moved as helper methods
AbstractDeviceParser::hasBrand(string brand)
AbstractDeviceParser::hasDeviceType(string deviceType)
private list:
parseResult(bool all = false) - this method implements new scanning logic
parseForBrand
shouldExistBrandOrException

Without indexing, tests are performed for 43 seconds, with 20 seconds.

Where can this be useful?
Mainly in long-lived processes, realtime services, background tasks, testing

**Comment by sanchezzzhak** [comment]
Rejected - this function is not needed yet.
Since there is memcache.

**I don't delete the branch in case it comes in handy.**

In Node Js, this thing works fine.

**Pull request closed without merging**